core/arch/arm/kernel/thread_spmc.c

```diff
@@ -1898,8 +1928,31 @@ struct mobj_ffa *thread_spmc_populate_mobj_from_rx(uint64_t cookie)
 	if (!mf)
 		goto out;
 
-	if (set_pages(descr->address_range_array,
-		      READ_ONCE(descr->address_range_count), num_pages, mf)) {
+	num_regions = MIN(READ_ONCE(descr->address_range_count),
+			  (my_rxtx.size - offs) /
+			  sizeof(struct ffa_address_range) - 1);
+	if (add_pages_at(descr->address_range_array, num_regions, &idx, mf)) {
+		mobj_ffa_spmc_delete(mf);
+		goto out;
+	}
+	/* Retrieve the remaining fragments. */
+	offs = frag_length;
+	handle = retrieve_desc.global_handle;
+	while (offs < length) {
+		buf = spmc_frag_retrieve(handle, offs, &frag_length);
+		if (buf) {
+			DMSG("Retrieve next fragment at offset %#x", offs);
+			num_regions = frag_length /
```

Comment:
Need to check that `frag_length` isn't larger than the size of the rx buffer.

Reply:
Done